govcd/extension_test.go

```diff
@@ -24,7 +26,112 @@ func (vcd *TestVCD) Test_GetExternalNetwork(check *C) {
+	check.Assert(err, IsNil)
+	if len(portGroups) == 0 {
+		check.Skip(fmt.Sprintf("No port group found with name '%s'", vcd.config.VCD.ExternalNetworkPortGroup))
+	}
+	if len(portGroups) > 1 {
+		check.Skip(fmt.Sprintf("More then one found with name '%s'", vcd.config.VCD.ExternalNetworkPortGroup))
+	}
+
+	externalNetwork := &types.ExternalNetwork{
+		Name:        TestCreateExternalNetwork,
+		Description: "Test Create External Network",
+		Xmlns:       types.XMLNamespaceExtension,
+		XmlnsVCloud: types.XMLNamespaceVCloud,
+		Configuration: &types.NetworkConfiguration{
+			Xmlns: types.XMLNamespaceVCloud,
+			IPScopes: &types.IPScopes{
+				IPScope: []*types.IPScope{&types.IPScope{
+					Gateway:   "192.168.201.1",
+					Netmask:   "255.255.255.0",
+					DNS1:      "192.168.202.253",
+					DNS2:      "192.168.202.254",
+					DNSSuffix: "some.net",
+					IPRanges: &types.IPRanges{
+						IPRange: []*types.IPRange{
+							&types.IPRange{
+								StartAddress: "192.168.201.3",
+								EndAddress:   "192.168.201.250",
+							},
+						},
+					},
+				},
+				}},
+			FenceMode: "isolated",
+		},
+		VimPortGroupRefs: &types.VimObjectRefs{
+			VimObjectRef: []*types.VimObjectRef{
+				&types.VimObjectRef{
+					VimServerRef: &types.Reference{
+						HREF: vimServerHref,
+					},
+					MoRef:         portGroups[0].MoRef,
+					VimObjectType: vcd.config.VCD.ExternalNetworkPortGroupType,
+				},
+			},
+		},
+	}
+	task, err := CreateExternalNetwork(vcd.client, externalNetwork)
+	check.Assert(err, IsNil)
+	check.Assert(task, Not(Equals), Task{})
+
+	err = task.WaitTaskCompletion()
+	check.Assert(err, IsNil)
+
+	newExternalNetwork := NewExternalNetwork(&vcd.client.Client)
+	err = newExternalNetwork.GetByName(TestCreateExternalNetwork)
+	check.Assert(err, IsNil)
+	check.Assert(newExternalNetwork.ExternalNetwork.Name, Equals, TestCreateExternalNetwork)
+
+	ipScope := newExternalNetwork.ExternalNetwork.Configuration.IPScopes.IPScope
+	check.Assert(ipScope[0].Gateway, Equals, "192.168.201.1")
+	check.Assert(ipScope[0].Netmask, Equals, "255.255.255.0")
+	check.Assert(ipScope[0].DNS1, Equals, "192.168.202.253")
+	check.Assert(ipScope[0].DNS2, Equals, "192.168.202.254")
+	check.Assert(ipScope[0].DNSSuffix, Equals, "some.net")
+
+	check.Assert(len(ipScope[0].IPRanges.IPRange), Equals, 1)
+	ipRange := ipScope[0].IPRanges.IPRange[0]
+	check.Assert(ipRange.StartAddress, Equals, "192.168.201.3")
+	check.Assert(ipRange.EndAddress, Equals, "192.168.201.250")
+
+	check.Assert(newExternalNetwork.ExternalNetwork.Configuration.FenceMode, Equals, "isolated")
+
+	// Needs to be deleted as port group used by other tests
+	// Workaround to refresh until task is fully completed - as task wait isn't enough
+	// Task still exists and creates NETWORK_DELETE error, so we wait until disappears
+	for i := 0; i < 30; i++ {
+		err = newExternalNetwork.Refresh()
+		check.Assert(err, IsNil)
+		if len(newExternalNetwork.ExternalNetwork.Tasks.Task) == 0 {
```

Comment:
What kind of task are we waiting to finish here?

Reply:
Creation task. I have spent many hours tracking the issue - seems in home lab, create task returns it's finished, but in reality it isn't for second or so, so had to add additional wait for checking if task finished.

Reply:
@vbauzysvmware what do you think with respect to @Didainius comment above? I'm concerned these lines in current test may be masking a potentially bigger issue.

Reply:
Race condition was due test and http handling go procedures. It is no implicate this case